<a href="https://colab.research.google.com/github/nabiila-29/data-fellowship-5/blob/main/Nabiila_Practice_Case_Statistics_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Step 1: Data Preprocessing

In [1]:
#import the library, read file from github, dataoverview
import pandas as pd                                                                                                  
df_raw1 = pd.read_csv("https://raw.githubusercontent.com/Syukrondzeko/Fellowship-5/main/Seasons_Stats.csv")          
df_raw1                                                                                                             

Unnamed: 0    Year             Player  Pos  ...   BLK   TOV     PF    PTS
0               0  1950.0    Curly Armstrong  G-F  ...   NaN   NaN  217.0  458.0
1               1  1950.0       Cliff Barker   SG  ...   NaN   NaN   99.0  279.0
2               2  1950.0      Leo Barnhorst   SF  ...   NaN   NaN  192.0  438.0
3               3  1950.0         Ed Bartels    F  ...   NaN   NaN   29.0   63.0
4               4  1950.0         Ed Bartels    F  ...   NaN   NaN   27.0   59.0
...           ...     ...                ...  ...  ...   ...   ...    ...    ...
24686       24686  2017.0        Cody Zeller   PF  ...  58.0  65.0  189.0  639.0
24687       24687  2017.0       Tyler Zeller    C  ...  21.0  20.0   61.0  178.0
24688       24688  2017.0  Stephen Zimmerman    C  ...   5.0   3.0   17.0   23.0
24689       24689  2017.0        Paul Zipser   SF  ...  16.0  40.0   78.0  240.0
24690       24690  2017.0        Ivica Zubac    C  ...  33.0  30.0   66.0  284.0

[24691 rows x 53 columns]

There are 24.691 rows and 53 column in this data type.

In [2]:
#filter data with year 2017 only
df_raw2 = df_raw1[df_raw1['Year'] == 2017]                         
df_raw2

Unnamed: 0    Year             Player Pos  ...   BLK    TOV     PF    PTS
24096       24096  2017.0       Alex Abrines  SG  ...   8.0   33.0  114.0  406.0
24097       24097  2017.0         Quincy Acy  PF  ...  15.0   21.0   67.0  222.0
24098       24098  2017.0         Quincy Acy  PF  ...   0.0    2.0    9.0   13.0
24099       24099  2017.0         Quincy Acy  PF  ...  15.0   19.0   58.0  209.0
24100       24100  2017.0       Steven Adams   C  ...  78.0  146.0  195.0  905.0
...           ...     ...                ...  ..  ...   ...    ...    ...    ...
24686       24686  2017.0        Cody Zeller  PF  ...  58.0   65.0  189.0  639.0
24687       24687  2017.0       Tyler Zeller   C  ...  21.0   20.0   61.0  178.0
24688       24688  2017.0  Stephen Zimmerman   C  ...   5.0    3.0   17.0   23.0
24689       24689  2017.0        Paul Zipser  SF  ...  16.0   40.0   78.0  240.0
24690       24690  2017.0        Ivica Zubac   C  ...  33.0   30.0   66.0  284.0

[595 rows x 53 columns]

The rows decreasing from 24.691 to 595 rows

In [3]:
#check data type
df_raw2.dtypes

Unnamed: 0      int64
Year          float64
Player         object
Pos            object
Age           float64
Tm             object
G             float64
GS            float64
MP            float64
PER           float64
TS%           float64
3PAr          float64
FTr           float64
ORB%          float64
DRB%          float64
TRB%          float64
AST%          float64
STL%          float64
BLK%          float64
TOV%          float64
USG%          float64
blanl         float64
OWS           float64
DWS           float64
WS            float64
WS/48         float64
blank2        float64
OBPM          float64
DBPM          float64
BPM           float64
VORP          float64
FG            float64
FGA           float64
FG%           float64
3P            float64
3PA           float64
3P%           float64
2P            float64
2PA           float64
2P%           float64
eFG%          float64
FT            float64
FTA           float64
FT%           float64
ORB           float64
DRB       

The date type has been match. We don’t need to do data transformation

In [4]:
#Check duplicate values
df_raw2.duplicated(['Player'])

24096    False
24097    False
24098     True
24099     True
24100    False
         ...  
24686    False
24687    False
24688    False
24689    False
24690    False
Length: 595, dtype: bool

'True' indicates there are duplicate values

In [5]:
#remove duplicate values
df_raw3 = df_raw2.drop_duplicates(subset=['Player'])  
df_raw3  

Unnamed: 0    Year             Player Pos  ...   BLK    TOV     PF    PTS
24096       24096  2017.0       Alex Abrines  SG  ...   8.0   33.0  114.0  406.0
24097       24097  2017.0         Quincy Acy  PF  ...  15.0   21.0   67.0  222.0
24100       24100  2017.0       Steven Adams   C  ...  78.0  146.0  195.0  905.0
24101       24101  2017.0      Arron Afflalo  SG  ...   7.0   42.0  104.0  515.0
24102       24102  2017.0      Alexis Ajinca   C  ...  22.0   31.0   77.0  207.0
...           ...     ...                ...  ..  ...   ...    ...    ...    ...
24686       24686  2017.0        Cody Zeller  PF  ...  58.0   65.0  189.0  639.0
24687       24687  2017.0       Tyler Zeller   C  ...  21.0   20.0   61.0  178.0
24688       24688  2017.0  Stephen Zimmerman   C  ...   5.0    3.0   17.0   23.0
24689       24689  2017.0        Paul Zipser  SF  ...  16.0   40.0   78.0  240.0
24690       24690  2017.0        Ivica Zubac   C  ...  33.0   30.0   66.0  284.0

[486 rows x 53 columns]

The rows decrease from 595 to 486 rows.

In [6]:
#check the null column
df_raw3.info()   

<class 'pandas.core.frame.DataFrame'>
Int64Index: 486 entries, 24096 to 24690
Data columns (total 53 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  486 non-null    int64  
 1   Year        486 non-null    float64
 2   Player      486 non-null    object 
 3   Pos         486 non-null    object 
 4   Age         486 non-null    float64
 5   Tm          486 non-null    object 
 6   G           486 non-null    float64
 7   GS          486 non-null    float64
 8   MP          486 non-null    float64
 9   PER         486 non-null    float64
 10  TS%         485 non-null    float64
 11  3PAr        485 non-null    float64
 12  FTr         485 non-null    float64
 13  ORB%        486 non-null    float64
 14  DRB%        486 non-null    float64
 15  TRB%        486 non-null    float64
 16  AST%        486 non-null    float64
 17  STL%        486 non-null    float64
 18  BLK%        486 non-null    float64
 19  TOV%        485 non-nul

0 indicates there is no value on the column. We will drop them.

In [7]:
#drop null columns
df = df_raw3.drop(columns=['blanl', 'blank2'])                
df

Unnamed: 0    Year             Player Pos  ...   BLK    TOV     PF    PTS
24096       24096  2017.0       Alex Abrines  SG  ...   8.0   33.0  114.0  406.0
24097       24097  2017.0         Quincy Acy  PF  ...  15.0   21.0   67.0  222.0
24100       24100  2017.0       Steven Adams   C  ...  78.0  146.0  195.0  905.0
24101       24101  2017.0      Arron Afflalo  SG  ...   7.0   42.0  104.0  515.0
24102       24102  2017.0      Alexis Ajinca   C  ...  22.0   31.0   77.0  207.0
...           ...     ...                ...  ..  ...   ...    ...    ...    ...
24686       24686  2017.0        Cody Zeller  PF  ...  58.0   65.0  189.0  639.0
24687       24687  2017.0       Tyler Zeller   C  ...  21.0   20.0   61.0  178.0
24688       24688  2017.0  Stephen Zimmerman   C  ...   5.0    3.0   17.0   23.0
24689       24689  2017.0        Paul Zipser  SF  ...  16.0   40.0   78.0  240.0
24690       24690  2017.0        Ivica Zubac   C  ...  33.0   30.0   66.0  284.0

[486 rows x 51 columns]

The columns decrease from 53 to 51

We have done our data reprocessing.
After this reprocessing, the data rows are 486 and columns is 51.

# Step 2: Answering Question

## 1. Who is the youngest and oldest player in the NBA in 2017 for each team (Tm) ?


In [8]:
#Chi-Square Normality test
from scipy import stats
import scipy as scipy
stats.chisquare(df['Age'])

Power_divergenceResult(statistic=346.79131925504555, pvalue=0.9999995782605277)

If the p-value> 0.05, the distribution of the data is normal. Above we have 0.9 **The data distribution is normal**

In [9]:
df[["Tm", "Player", "Age"]].head() 

Tm         Player   Age
24096  OKC   Alex Abrines  23.0
24097  TOT     Quincy Acy  26.0
24100  OKC   Steven Adams  23.0
24101  SAC  Arron Afflalo  31.0
24102  NOP  Alexis Ajinca  28.0

In [10]:
#fing the youngest player on every team
youngest = df.groupby('Tm')['Age', 'Player'].agg(min)
youngest

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  


Age            Player
Tm                         
ATL  22.0   DeAndre' Bembry
BOS  20.0        Al Horford
BRK  21.0   Anthony Bennett
CHI  21.0      Bobby Portis
CHO  21.0    Aaron Harrison
CLE  21.0     Channing Frye
DAL  21.0      A.J. Hammons
DEN  19.0        Alonzo Gee
DET  20.0    Andre Drummond
GSW  20.0  Anderson Varejao
HOU  20.0       Bobby Brown
IND  20.0      Aaron Brooks
LAC  19.0     Alan Anderson
LAL  19.0    Brandon Ingram
MEM  20.0   Andrew Harrison
MIA  20.0      Dion Waiters
MIL  19.0       Gary Payton
MIN  20.0     Adreian Payne
NOP  20.0     Alexis Ajinca
NYK  21.0   Carmelo Anthony
OKC  20.0      Alex Abrines
ORL  20.0      Aaron Gordon
PHI  21.0    Alex Poythress
PHO  19.0     Alan Williams
POR  21.0   Al-Farouq Aminu
SAC  19.0  Anthony Tolliver
SAS  20.0       Bryn Forbes
TOR  21.0     Bruno Caboclo
TOT  21.0      Andrew Bogut
UTA  21.0        Alec Burks
WAS  21.0      Bradley Beal

In [11]:
#find the oldest player on every team.
oldest = df.groupby('Tm')['Age', 'Player'].agg(max)
oldest

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  


Age                   Player
Tm                                
ATL  32.0             Tim Hardaway
BOS  31.0             Tyler Zeller
BRK  36.0            Trevor Booker
CHI  35.0              Robin Lopez
CHO  31.0           Treveon Graham
CLE  38.0         Tristan Thompson
DAL  38.0          Wesley Matthews
DEN  36.0          Wilson Chandler
DET  34.0            Tobias Harris
GSW  36.0            Zaza Pachulia
HOU  34.0             Trevor Ariza
IND  32.0           Thaddeus Young
LAC  39.0           Wesley Johnson
LAL  37.0           Timofey Mozgov
MEM  40.0            Zach Randolph
MIA  36.0              Willie Reed
MIL  39.0               Tony Snell
MIN  34.0              Zach LaVine
NOP  33.0              Tim Frazier
NYK  32.0        Willy Hernangomez
OKC  36.0           Victor Oladipo
ORL  32.0        Stephen Zimmerman
PHI  32.0  Timothe Luwawu-Cabarrot
PHO  34.0           Tyson Chandler
POR  28.0           Tim Quarterman
SAC  31.0      Willie Cauley-Stein
SAS  39.0              Tony Parker
TOR  30.0        Patrick Patterson
TOT  36.0             Yogi Ferrell
UTA  35.0               Trey Lyles
WAS  32.0               Trey Burke

In [12]:
#optinal. Join the data
no1 = pd.merge(youngest, oldest, on='Tm', how = 'inner')
no1.head()

Age_x         Player_x  Age_y        Player_y
Tm                                                
ATL   22.0  DeAndre' Bembry   32.0    Tim Hardaway
BOS   20.0       Al Horford   31.0    Tyler Zeller
BRK   21.0  Anthony Bennett   36.0   Trevor Booker
CHI   21.0     Bobby Portis   35.0     Robin Lopez
CHO   21.0   Aaron Harrison   31.0  Treveon Graham

In [13]:
#rename the column name
no1.rename(columns ={'Age_x' : 'YoungestAge', 'Age_y' : 'OldestAge', 'Player_x': 'YoungestPlayer', 'Player_y' : 'OldestPlayer'}, inplace=True )
no1.head()

YoungestAge   YoungestPlayer  OldestAge    OldestPlayer
Tm                                                          
ATL         22.0  DeAndre' Bembry       32.0    Tim Hardaway
BOS         20.0       Al Horford       31.0    Tyler Zeller
BRK         21.0  Anthony Bennett       36.0   Trevor Booker
CHI         21.0     Bobby Portis       35.0     Robin Lopez
CHO         21.0   Aaron Harrison       31.0  Treveon Graham

## 2. Which player has the most minutes played (MP) in each position (Pos)?

In [14]:
#shapiro vilk test (normality test)
stats.shapiro(df['MP'])

(0.9472718238830566, 3.893537402210878e-12)

We see the p-value. The p-value is the second index of the result above. The normal data has p-value 0> 0.05

the p-value of the data is 3.89. **The data has normal distirbution**

In [15]:
#player with most minutes played on every position
no2 = df.groupby('Pos')['MP', 'Player'].agg(max)
no2

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  


MP             Player
Pos                            
C     3030.0      Zaza Pachulia
PF    2803.0      Zach Randolph
PF-C   980.0  Joffrey Lauvergne
PG    2947.0       Yogi Ferrell
SF    3048.0    Wilson Chandler
SG    2796.0        Zach LaVine

## 3. Which team has the highest average total rebound percentage (TRB%), assist percentage (AST%), steal percentage (STL%), and block percentage (BLK%)?


### TRB%

In [16]:
#team with highest average of TRB%
trb = df.groupby(['Tm']).mean()['TRB%'].sort_values(ascending=False)
trb.head(1)

Tm
WAS    13.45
Name: TRB%, dtype: float64

Team with highest TRB% is WAS

### AST%

In [17]:
#team with highest average of AST%
ast = df.groupby(['Tm']).mean()['AST%'].sort_values(ascending=False)
ast.head(1)

Tm
DEN    15.86
Name: AST%, dtype: float64

Team with highest AST% is DEN

### STL%

In [18]:
#team with highest average of STL%
stl = df.groupby(['Tm']).mean()['STL%'].sort_values(ascending=False)
stl.head(1)

Tm
MIN    2.371429
Name: STL%, dtype: float64

### BLK%

In [19]:
#Team with highest average of BLK%
blk = df.groupby(['Tm']).mean()['BLK%'].sort_values(ascending=False)
blk.head(1)

Tm
GSW    2.74
Name: BLK%, dtype: float64

Team with highest BLK% is GSW

## 4. Who is the best player in your opinion based on his record stats? note: you can refer to variables point (PTS), assists, rebounds, or anything else. A combination of several variables would be nice. 

Here, we will use WS as the indicator of best player

In [20]:
#best player according to Win Share
best_player= df.sort_values('WS', ascending=False)
best_player[['Player', 'WS']].head(2)


Player    WS
24306  James Harden  15.0
24285   Rudy Gobert  14.3

James Harden is the best player according to WS values

## 5. Which team has the best average stat record of their players? Note: you can refer to points,assists, rebounds, or anything else. A combination of several variables would be nice 

In [21]:
#best team according to Win Share 
best_team = df.groupby(['Tm']).mean()['WS'].sort_values(ascending=False)
best_team.head(2)

Tm
GSW    4.460000
HOU    4.061538
Name: WS, dtype: float64

We use WS as the indicator of the best stat record. We group it by team. And found that GSW is the best team.